File: Java/0029_divide_two_integers/0029_divide_two_integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        int sign = (dividend > 0) ^ (divisor > 0) ? -1 : 1;
        long res = 0;
        long m = Math.abs((long)dividend);
        long n = Math.abs((long)divisor);
        while (m >= n) {
            long temp = n;
            long i = 1;
            while (temp << 1 < m) {
                temp <<= 1;
                i <<= 1;
            }
            m -= temp;
            res += i;
        }
        if (sign < 0)
            res = -res;
        return res > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int)res;
    }
}

File: Python/0029_divide_two_integers/0029_divide_two_integers.py
class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        sign = -1 if (1 if dividend > 0 else 0) ^ (1 if divisor > 0 else 0) > 0 else 1
        res = 0
        m = abs(dividend)
        n = abs(divisor)
        while m >= n:
            temp, i = n, 1
            while (temp << 1) < m:
                temp <<= 1
                i <<= 1
            m -= temp
            res += i
        if sign < 0:
            res = -res
        return res if res <= 2147483647 else 2147483647
